src/KubernetesClient/Kubernetes.Generic.cs

```diff
@@ -0,0 +1,1131 @@
+            string labelSelector = default,
+            int? limit = default,
+            string resourceVersion = default,
+            int? timeoutSeconds = default,
+            bool? watch = default,
+            string pretty = default,
+            Dictionary<string, List<string>> customHeaders = null,
+            CancellationToken cancellationToken = default)
+            where T : IKubernetesObject
+        {
+            var listType = typeof(KubernetesList<>).MakeGenericType(typeof(T));
+            var result = await ListWithHttpMessagesAsync(listType, namespaceParameter, allowWatchBookmarks, continueParameter, fieldSelector, labelSelector, limit, resourceVersion, timeoutSeconds, watch, pretty, customHeaders, cancellationToken);
+            return (HttpOperationResponse<KubernetesList<T>>) result;
+        }
+
+        /// <inheritdoc cref="IKubernetes"/>
+        public async Task<HttpOperationResponse> ListWithHttpMessagesAsync(
+            Type type,
+            string namespaceParameter = default,
+            bool? allowWatchBookmarks = default,
+            string continueParameter = default,
+            string fieldSelector = default,
+            string labelSelector = default,
+            int? limit = default,
+            string resourceVersion = default,
+            int? timeoutSeconds = default,
+            bool? watch = default,
+            string pretty = default,
+            Dictionary<string, List<string>> customHeaders = null,
+            CancellationToken cancellationToken = default)
+        {
+
+
+            // Tracing
+            var shouldTrace = ServiceClientTracing.IsEnabled;
+            string invocationId = null;
+            if (shouldTrace)
+            {
+                invocationId = ServiceClientTracing.NextInvocationId.ToString();
+                var tracingParameters = new Dictionary<string, object>();
+                tracingParameters.Add("allowWatchBookmarks", allowWatchBookmarks);
+                tracingParameters.Add("continueParameter", continueParameter);
+                tracingParameters.Add("fieldSelector", fieldSelector);
+                tracingParameters.Add("labelSelector", labelSelector);
+                tracingParameters.Add("limit", limit);
+                tracingParameters.Add("resourceVersion", resourceVersion);
+                tracingParameters.Add("timeoutSeconds", timeoutSeconds);
+                tracingParameters.Add("watch", watch);
+                tracingParameters.Add("namespaceParameter", namespaceParameter);
+                tracingParameters.Add("pretty", pretty);
+                tracingParameters.Add("cancellationToken", cancellationToken);
+                ServiceClientTracing.Enter(invocationId, this, "ListNamespacedPod", tracingParameters);
+            }
+            var url = GetUrlPathForResource(type, namespaceParameter);
+            var queryParameters = new List<string>();
+            if (allowWatchBookmarks != null)
```

Comment:
There is logic for building query strings here:

https://github.com/kubernetes-client/csharp/blob/master/src/KubernetesClient/Utilities.cs#L9

That will clean this up a lot.